File: ClaudeSync/ClaudeSync/Security/TrustStore.swift
// TrustStore.swift
// ClaudeSync
//
// Persists paired device information to ~/Library/Application Support/claude-sync/paired_devices.json.
// Stores certificate fingerprints (SHA-256 of DER-encoded cert) so that TLS connections
// from paired devices can be verified without a central CA.
//
// All mutable state is actor-isolated for thread safety. Disk I/O uses atomic writes
// to prevent corruption from concurrent access or crashes.

import Foundation
import os

/// Persists paired device trust information.
/// Stores certificate fingerprints keyed by device ID in a JSON file.
actor TrustStore {

    // MARK: - Types

    /// A single trusted device record.
    struct TrustedDevice: Codable, Sendable {
        /// The peer's persistent device UUID.
        let deviceId: String

        /// Human-readable device name at pairing time.
        let name: String

        /// SHA-256 fingerprint of the peer's DER-encoded certificate.
        let certFingerprint: String

        /// Timestamp when the pairing was established.
        let pairedAt: Date

        private enum CodingKeys: String, CodingKey {
            case deviceId = "device_id"
            case name
            case certFingerprint = "cert_fingerprint"
            case pairedAt = "paired_at"
        }
    }

    // MARK: - Properties

    /// Path to the trust store JSON file.
    private let storePath: URL

    /// In-memory cache of trusted devices, keyed by device ID.
    private var trustedDevices: [String: TrustedDevice] = [:]

    /// Whether the trust store has been loaded from disk.
    private var isLoaded: Bool = false

    /// Logger for trust store operations.
    private let logger = Logger(subsystem: "com.claudesync", category: "TrustStore")

    /// JSON encoder configured with ISO 8601 dates and sorted keys for deterministic output.
    private static let encoder: JSONEncoder = {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        return encoder
    }()

    /// JSON decoder configured to match the encoder's date strategy.
    private static let decoder: JSONDecoder = {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return decoder
    }()

    // MARK: - Initialization

    /// Creates a trust store at the default location.
    /// - Parameter storePath: Override path for testing. Defaults to
    ///   ~/Library/Application Support/claude-sync/paired_devices.json
    init(storePath: URL? = nil) {
        if let storePath = storePath {
            self.storePath = storePath
        } else {
            let appSupport = FileManager.default.urls(
                for: .applicationSupportDirectory,
                in: .userDomainMask
            ).first!
            self.storePath = appSupport
                .appendingPathComponent("claude-sync")
                .appendingPathComponent("paired_devices.json")
        }
    }

    // MARK: - Load / Save

    /// Loads the trust store from disk. Safe to call multiple times (idempotent after first load).
    func load() async throws {
        guard !isLoaded else { return }

        let fm = FileManager.default
        guard fm.fileExists(atPath: storePath.path) else {
            // No store file yet -- start with empty trust set.
            isLoaded = true
            logger.info("No paired_devices.json found, starting with empty trust store")
            return
        }

        let data = try Data(contentsOf: storePath)
        let devices = try Self.decoder.decode([TrustedDevice].self, from: data)

        trustedDevices = Dictionary(uniqueKeysWithValues: devices.map { ($0.deviceId, $0) })
        isLoaded = true
        logger.info("Loaded trust store: \(devices.count) paired devices")
    }

    /// Saves the current trust store to disk atomically.
    func save() async throws {
        let devices = Array(trustedDevices.values)
            .sorted { $0.pairedAt < $1.pairedAt } // Deterministic order by pairing time.

        let data = try Self.encoder.encode(devices)

        // Ensure the parent directory exists.
        let parentDir = storePath.deletingLastPathComponent()
        try FileManager.default.createDirectory(at: parentDir, withIntermediateDirectories: true)

        // Write atomically to prevent corruption.
        try data.write(to: storePath, options: .atomic)
        logger.info("Saved trust store: \(devices.count) paired devices")
    }

    // MARK: - Trust Management

    /// Adds a trusted device to the store and persists to disk.
    /// If a device with the same ID already exists, it is replaced.
    func addTrustedDevice(_ device: TrustedDevice) async throws {
        if !isLoaded { try await load() }

        trustedDevices[device.deviceId] = device
        try await save()
        logger.info("Added trusted device: \(device.name) (\(device.deviceId.prefix(8)))")
    }

    /// Removes a trusted device from the store and persists to disk.
    func removeTrustedDevice(deviceId: String) async throws {
        if !isLoaded { try await load() }

        let removed = trustedDevices.removeValue(forKey: deviceId)
        if removed != nil {
            try await save()
            logger.info("Removed trusted device: \(deviceId.prefix(8))")
        }
    }

    /// Checks if a device is trusted by its device ID.
    func isTrusted(deviceId: String) -> Bool {
        return trustedDevices[deviceId] != nil
    }

    /// Checks if a certificate fingerprint belongs to any trusted device.
    func isTrustedCert(fingerprint: String) -> Bool {
        return trustedDevices.values.contains { $0.certFingerprint == fingerprint }
    }

    /// Returns the trusted device record for a given device ID, if it exists.
    func trustedDevice(for deviceId: String) -> TrustedDevice? {
        return trustedDevices[deviceId]
    }

    /// Returns all trusted devices sorted by pairing date.
    func allTrustedDevices() -> [TrustedDevice] {
        return Array(trustedDevices.values).sorted { $0.pairedAt < $1.pairedAt }
    }

    /// Returns the total number of trusted devices.
    func count() -> Int {
        return trustedDevices.count
    }
}

File: claude-sync-app/src-tauri/src/security/trust_store.rs
// ==========================================================================
// Trust store - persists paired device information to disk
// Stored at ~/Library/Application Support/claude-sync/paired_devices.json
// (macOS) or equivalent on other platforms. Thread-safe read/write.
// ==========================================================================

use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;

/// File name for the persisted trust store.
const TRUST_STORE_FILENAME: &str = "paired_devices.json";

/// A single trusted (paired) device record.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TrustedDevice {
    pub device_id: String,
    pub name: String,
    pub cert_fingerprint: String,
    /// When the device was paired (Unix epoch seconds)
    pub paired_at: i64,
}

/// Serialization wrapper for the trust store file.
#[derive(Debug, Clone, Serialize, Deserialize)]
struct TrustStoreData {
    devices: HashMap<String, TrustedDevice>,
}

/// Persists paired device trust information to disk.
/// Provides add/remove/query operations with automatic persistence.
pub struct TrustStore {
    /// Path to the directory containing the trust store file
    store_path: PathBuf,
    /// In-memory representation of trusted devices
    data: TrustStoreData,
}

impl TrustStore {
    /// Create a new TrustStore using the platform data directory.
    /// Call `load()` after construction to read existing data from disk.
    pub fn new() -> Self {
        let store_path = dirs::data_dir()
            .unwrap_or_else(|| PathBuf::from("."))
            .join("claude-sync");
        Self {
            store_path,
            data: TrustStoreData {
                devices: HashMap::new(),
            },
        }
    }

    /// Load existing trust data from disk. If the file doesn't exist,
    /// the store starts empty (not an error).
    pub fn load(&mut self) -> Result<(), Box<dyn std::error::Error>> {
        let file_path = self.store_path.join(TRUST_STORE_FILENAME);
        if !file_path.exists() {
            log::info!("No trust store file found at {:?}, starting empty", file_path);
            return Ok(());
        }

        let contents = fs::read_to_string(&file_path)?;
        self.data = serde_json::from_str(&contents)?;
        log::info!(
            "Loaded trust store with {} paired devices",
            self.data.devices.len()
        );
        Ok(())
    }

    /// Save the current trust data to disk atomically (write to temp, then rename).
    pub fn save(&self) -> Result<(), Box<dyn std::error::Error>> {
        fs::create_dir_all(&self.store_path)?;

        let file_path = self.store_path.join(TRUST_STORE_FILENAME);
        let tmp_path = file_path.with_extension("json.tmp");

        let json = serde_json::to_string_pretty(&self.data)?;
        fs::write(&tmp_path, &json)?;
        fs::rename(&tmp_path, &file_path)?;

        Ok(())
    }

    /// Add a device to the trust store and persist to disk.
    pub fn add_trusted_device(&mut self, device: TrustedDevice) -> Result<(), Box<dyn std::error::Error>> {
        log::info!(
            "Adding trusted device: {} ({})",
            device.name,
            device.device_id
        );
        self.data.devices.insert(device.device_id.clone(), device);
        self.save()
    }

    /// Remove a device from the trust store and persist the change.
    pub fn remove_trusted_device(&mut self, device_id: &str) -> Result<(), Box<dyn std::error::Error>> {
        if self.data.devices.remove(device_id).is_some() {
            log::info!("Removed trusted device: {}", device_id);
            self.save()?;
        }
        Ok(())
    }

    /// Check if a device is trusted by its device_id.
    pub fn is_trusted(&self, device_id: &str) -> bool {
        self.data.devices.contains_key(device_id)
    }

    /// Check if a certificate fingerprint belongs to a trusted device.
    pub fn is_trusted_cert(&self, fingerprint: &str) -> bool {
        self.data
            .devices
            .values()
            .any(|d| d.cert_fingerprint == fingerprint)
    }

    /// Get all trusted devices as a list.
    pub fn all_trusted_devices(&self) -> Vec<TrustedDevice> {
        self.data.devices.values().cloned().collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_trust_store_in_memory() {
        let mut store = TrustStore {
            store_path: PathBuf::from("/tmp/test-claude-sync-trust"),
            data: TrustStoreData {
                devices: HashMap::new(),
            },
        };

        assert!(!store.is_trusted("device-1"));
        assert!(!store.is_trusted_cert("AA:BB:CC"));

        let device = TrustedDevice {
            device_id: "device-1".to_string(),
            name: "Test MacBook".to_string(),
            cert_fingerprint: "AA:BB:CC".to_string(),
            paired_at: 1700000000,
        };

        // Don't persist in test; just modify in-memory
        store.data.devices.insert(device.device_id.clone(), device);

        assert!(store.is_trusted("device-1"));
        assert!(store.is_trusted_cert("AA:BB:CC"));
        assert!(!store.is_trusted("device-2"));

        assert_eq!(store.all_trusted_devices().len(), 1);
    }
}
